tests/e2e/etcd_mix_versions_test.go

```diff
@@ -123,3 +112,102 @@ func mixVersionsSnapshotTest(t *testing.T, clusterVersion, newInstanceVersion e2
 		return true
 	}, 10*time.Second, 500*time.Millisecond)
 }
+
+func TestMixVersionsSnapshotByMockingPartition(t *testing.T) {
+	cases := []struct {
+		name                   string
+		clusterVersion         e2e.ClusterVersion
+		mockPartitionNodeIndex int
+	}{
+		{
+			name:                   "etcd instance with previous version receives snapshot from instance with current version",
+			clusterVersion:         e2e.MinorityLastVersion,
+			mockPartitionNodeIndex: 0,
+		},
+		{
+			name:                   "etcd instance with current version receives snapshot from instance with previous version",
+			clusterVersion:         e2e.MinorityLastVersion,
+			mockPartitionNodeIndex: 2,
+		},
+	}
+
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			mixVersionsSnapshotTestByMockPartition(t, tc.clusterVersion, tc.mockPartitionNodeIndex)
+		})
+	}
+}
+
+func mixVersionsSnapshotTestByMockPartition(t *testing.T, clusterVersion e2e.ClusterVersion, mockPartitionNodeIndex int) {
+	e2e.BeforeTest(t)
+
+	if !fileutil.Exist(e2e.BinPath.EtcdLastRelease) {
+		t.Skipf("%q does not exist", e2e.BinPath.EtcdLastRelease)
+	}
+
+	// Create an etcd cluster with 3 member of MinorityLastVersion
+	epc, err := e2e.NewEtcdProcessCluster(context.TODO(), t,
+		e2e.WithClusterSize(3),
+		e2e.WithSnapshotCount(10),
+		e2e.WithVersion(clusterVersion),
+	)
+	if err != nil {
+		t.Fatalf("failed to start etcd cluster: %v", err)
+	}
+	defer func() {
+		if err := epc.Close(); err != nil {
+			t.Fatalf("failed to close etcd cluster: %v", err)
+		}
+	}()
+	toPartitionedMember := epc.Procs[mockPartitionNodeIndex]
+
+	// Stop and restart the partitioned member
+	if err = toPartitionedMember.Stop(); err != nil {
+		t.Fatal(err)
+	}
+
+	// Write more than SnapshotCount entries to trigger at least a snapshot.
+	t.Log("Writing 20 keys to the cluster")
```

Comment:
Have you double confirmed a snapshot was indeed sent out by checking the test log? I am afraid you need to write at least 5K entries in this case.  

```
DefaultSnapshotCatchUpEntries uint64 = 5000
```

Reply:
FYI. https://github.com/etcd-io/etcd/pull/14819#discussion_r1052917273

Reply:
Sorry I don't really get it,  I already set `SnapshotCount` to 10 in the e2e case, should  `DefaultSnapshotCatchUpEntries` be concerned?

And cannot find the etcd server logs when I ran e2e test cases (only see the e2e client logs),  where should I find it ?  Thanks

Reply:
Hi, I checkout the server's log, and found that the snapshot is saved by the leader,  while snapshot is not sent from leader to follower when follower restarts again.

<img width="2078" alt="Screen Shot 2023-01-02 at 11 30 12 PM" src="https://user-images.githubusercontent.com/13792102/210251797-1cf767f1-47d1-4049-ab24-19fd192d03e2.png">

And I go through the raft code, and I was suspecting it didn't go to the `if errt != nil || erre != nil `  statement  in the following code block.
while I check the e2e test case logs, and be sure that follower is restarted right after leader has saved snapshot , and don't know why it don't get here? 

![image](https://user-images.githubusercontent.com/13792102/210252141-4279d54b-62db-44b7-bf5d-5f4c2de066b0.png)

Reply:
It means the leader did not send snapshot to the follower at all. The snapshot mentioned in your first screen shot is just V2 snapshot, it isn't the same thing as the snapshot what we are talking about.  The snapshot sent by the leader to follower is actually a complete copy of the boltDB file. 

FYI. [etcd persistent storage files](https://docs.google.com/document/d/1O2o1IApHWmSioXG3fez4eVlUHOrXICYGNVIzaqNS0IQ/edit?resourcekey=0-e6Iywgdkol0uiVBAaV1oww#heading=h.xye8zm7sz9wr )

After https://github.com/etcd-io/etcd/pull/15033 is merged, you can set a small value for `--experimental-snapshot-catch-up-entries` as well.

Reply:
OK, it seems due to large SnapshotCatchUpEntries value, it not trigger `s.r.raftStorage.Compact(compacti)`,  so it not trigger leader send snapshot to follower, I will wait https://github.com/etcd-io/etcd/pull/15033 merged and try again.

Reply:
The PR https://github.com/etcd-io/etcd/pull/15033 has been merged. Please rebase & continue to work this PR

Reply:
> The PR #15033 has been merged. Please rebase & continue to work this PR

hi , EtcdLastRelease (v3.5) binary not contains https://github.com/etcd-io/etcd/pull/15033 changes, should https://github.com/etcd-io/etcd/pull/15033 be picked to v3.5 too ?

Reply:
The case of `3.5 sending snapshot to 3.6` has already been covered in the test case `TestMixVersionsSnapshotByAddingMember`, so you can cover the case of `3.6 sending snapshot to 3.5` in the new test case. 

I think it's a good case to verify both mix-version snapshot and whether https://github.com/etcd-io/etcd/pull/15033 has impact on mix-version scenario.

Reply:
@ahrtr hi, I've change my PR and set `SnapshotCatchUpEntries` to 10 to trigger leader sending snapshot to follower, and I can verify it worked correctly by etcd logs, please have a review again, thanks !